ostd/src/mm/frame/allocator.rs

```diff
@@ -140,73 +125,207 @@ fn test_alloc_dealloc() {
+                max_u4g_size = region.len();
+                max_u4g_index = i;
+            }
+            if region.base() >= PADDR4G && region.len() > max_size {
+                max_size = region.len();
+                max_index = i;
+            }
+        }
+        let under_4g_region = regions[max_u4g_index];
+        let max_region = (max_index != 0).then(|| regions[max_index]);
+        log::debug!(
+            "Early frame allocator (below 4G) at: {:#x}-{:#x}",
+            under_4g_region.base(),
+            under_4g_region.end()
+        );
+        if max_region.is_some() {
+            log::debug!(
+                "Early frame allocator (above 4G) at: {:#x}-{:#x}",
+                max_region.as_ref().unwrap().base(),
+                max_region.as_ref().unwrap().end()
             );
         }
+        let under_4g_end = under_4g_region.base().align_up(PAGE_SIZE);
+        let max_end = max_region
+            .as_ref()
+            .map(|region| region.base().align_up(PAGE_SIZE));
+        Self {
+            under_4g_region,
+            under_4g_end,
+            max_region,
+            max_end,
+        }
+    }
+
+    /// Allocates a contiguous range of frames.
+    pub fn alloc(&mut self, layout: Layout) -> Option<Paddr> {
+        let size = layout.size().align_up(PAGE_SIZE);
+        let allocated = self.under_4g_end.align_up(layout.align());
+        if allocated + size <= self.under_4g_region.end() {
+            // Allocated below 4G.
+            self.under_4g_end = allocated + size;
+            Some(allocated)
+        } else {
+            // Try above 4G.
+            let max_end = self.max_end.as_mut()?;
+            let allocated = max_end.align_up(layout.align());
+            if allocated + size > self.max_region.as_ref().unwrap().end() {
+                return None;
+            }
+            *max_end = allocated + size;
+            Some(allocated)
+        }
+    }
+
+    pub(super) fn allocated_regions(&self) -> (Range<Paddr>, Option<Range<Paddr>>) {
+        (
+            self.under_4g_region.base().align_up(PAGE_SIZE)..self.under_4g_end,
+            self.max_region
+                .as_ref()
+                .map(|region| region.base().align_up(PAGE_SIZE)..self.max_end.unwrap()),
+        )
+    }
+}
+
+/// Metadata for frames allocated in the early boot phase.
+///
+/// Frames allocated with [`early_alloc`] are not immediately tracked with
+/// frame metadata. But [`super::meta::init`] will track them later.
+#[derive(Debug)]
+pub(crate) struct EarlyAllocatedFrameMeta;
+
+impl_frame_meta_for!(EarlyAllocatedFrameMeta);
+
+/// Allocates a contiguous range of frames in the early boot phase.
+///
+/// The early allocated frames will not be reclaimable, until the metadata is
+/// initialized by [`super::meta::init`]. Then we can use [`Frame::from_raw`]
+/// to free the frames.
+pub(crate) fn early_alloc(layout: Layout) -> Option<Paddr> {
```

Comment:
This is not safe if the memory map is not ready. So perhaps add another `init_early_alloc` method and call it at the appropriate time.
```rust
/// # Safety
/// The caller must make sure that the memory map is ready, i.e. that all free regions specified in
/// the memory map are really free and do not overlap with kernel code and kernel data.
unsafe fn init_early_alloc() {}
```

This also makes the above suggestion of changing `allocated_regions` to `into_allocated_regions` work, so that the early allocator doesn't accidentally get reinitialized after the metadata initialization.